### PR TITLE
test: Add stress tests for OCI and Helm

### PR DESCRIPTION
**Pull request by karlkfi**
- Add TestStressManyDeployments to pre-load stress test clusters,
  so that the control plane auto-scales before using it for other
  stress tests.
- Add TestStressMemoryUsageHelm to ensure the helm-sync container
  works at scale.
- Add TestStressMemoryUsageOCI to ensure the oci-sync container
  works at scale.

Dependencies:
- https://github.com/kptdev/kpt/pull/4055
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/937
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/928
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/945
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/947
- https://github.com/GoogleContainerTools/kpt-config-sync/pull/948

Related Profiling: go/config-sync-reconciler-profiling-2

### CODE REVIEW

**Comment by karlkfi** [comment]
/hold

Blocked on https://github.com/kptdev/kpt/pull/4055 and need to push gke-e2e

File: anvil-set.yaml
Comment (nan-yu) [inline anvil-set.yaml:1]: Is this file needed?

File: e2e/nomostest/artifactregistry/crane.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		return fmt.Errorf("packaging image: %w", err)
+	}
+	r.Logger.Infof("Pushing image: %s:%s", r.Name, r.Version)
+	_, err := r.Shell.ExecWithDebug("crane", "append",
+		"-f", localSourceTgzPath,
+		"-t", fmt.Sprintf("%s:%s", r.Address(), r.Version))
+	if err != nil {
+		return fmt.Errorf("pushing image: %w", err)
+	}
+	if err := os.Remove(localSourceTgzPath); err != nil {
```
Comment (nan-yu) [inline e2e/nomostest/artifactregistry/crane.go:86]: Suggest moving into a defer block so it can be deleted if the push fails.

File: e2e/nomostest/artifactregistry/helm.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	parentPath := filepath.Dir(r.Image.BuildPath)
+	if _, err := r.Image.Shell.Helm("package", r.Image.BuildPath, "--destination", parentPath+string(filepath.Separator)); err != nil {
+		return fmt.Errorf("packaging helm chart: %w", err)
+	}
+	r.Image.Logger.Infof("Pushing helm chart: %s:%s", r.Image.Name, r.Image.Version)
+	chartFile := filepath.Join(parentPath, fmt.Sprintf("%s-%s.tgz", r.Image.Name, r.Image.Version))
+	if _, err := r.Image.Shell.Helm("push", chartFile, r.Image.RepositoryOCI()); err != nil {
+		return fmt.Errorf("pushing helm chart: %w", err)
+	}
+	if err := os.Remove(chartFile); err != nil {
```
Comment (nan-yu) [inline e2e/nomostest/artifactregistry/helm.go:159]: ditto

File: e2e/nomostest/iam/account.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"fmt"
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest"
+)
+
+// ValidateServiceAccountExists validates that the specified GSA exists in the
+// test project.
+func ValidateServiceAccountExists(nt *nomostest.NT, gsaEmail string) error {
```
Comment (nan-yu) [inline e2e/nomostest/iam/account.go:26]: We can also validate if the SA has the expected access.

File: e2e/testcases/stress_test.go [hunk 4]
```diff
@@ -332,11 +307,89 @@ func TestStress100CRDs(t *testing.T) {
 	}
 }
 
-// TestStressMemoryUsage applies 100 CRDs and then 10 objects for each resource.
-// This stressed both the number of watches and the number of objects,
+// TestStressManyDeployments applies 1000 Deployments with 1 replica each.
+// This ensures that Config Sync can apply resources while the cluster control
```
Comment (nan-yu) [inline e2e/testcases/stress_test.go:311]: 100 or 1000?

File: e2e/nomostest/artifactregistry/crane.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// CraneImage represents a remote OCI-based crane-build image
+type CraneImage struct {
+	*Image
+}
+
+// RegistryLogin will log into the registry with crane using a gcloud auth token.
+func (r *CraneImage) RegistryLogin() error {
+	var err error
+	authCmd := r.Shell.Command("gcloud", "auth", "print-access-token")
+	loginCmd := r.Shell.Command("crane", "auth", "login",
```
Comment (nan-yu) [inline e2e/nomostest/artifactregistry/crane.go:51]: You chose to use `krane` in your other PR. What's the reason to switch to `crane`?

File: e2e/nomostest/artifactregistry/crane.go (hunk 1 above)
Comment (karlkfi) [reply e2e/nomostest/artifactregistry/crane.go:86]: The whole TmpDir is deleted after the test. So it's not needed. This just helps conserve space and avoid hypothetical error if the same chart is built twice.

File: e2e/nomostest/iam/account.go (hunk 3 above)
Comment (karlkfi) [reply e2e/nomostest/iam/account.go:26]: I'm not entirely sure how to do that. Would you like to do that in a follow up PR?

File: e2e/testcases/stress_test.go (hunk 4 above)
Comment (karlkfi) [reply e2e/testcases/stress_test.go:311]: oops. I set it to 100 for a quick test and forgot to set it back. I'm running 1000 right now locall to check that it still works.

File: e2e/nomostest/artifactregistry/crane.go (hunk 5 above)
Comment (karlkfi) [reply e2e/nomostest/artifactregistry/crane.go:51]: That PR wasn't merged. @sdowell suggested another way to do it and made a different PR to replace it. So the current buildenv included crane but not krane.

Since our docs use crane, I stuck with that. The only think krane does is handle auth for you so you don't need to login or configure gcloud helper in the docker config.

File: anvil-set.yaml
Comment (karlkfi) [reply anvil-set.yaml:1]: nope. will remove

File: e2e/nomostest/iam/account.go (hunk 3 above)
Comment (karlkfi) [reply e2e/nomostest/iam/account.go:26]: FWIW, I didn't add this check, just refactored it into a function, since it was being used multiple places.

File: anvil-set.yaml
Comment (karlkfi) [reply anvil-set.yaml:1]: removed

File: e2e/nomostest/artifactregistry/crane.go (hunk 1 above)
Comment (karlkfi) [reply e2e/nomostest/artifactregistry/crane.go:88]: Added a comment to explain

File: e2e/testcases/stress_test.go (hunk 4 above)
Comment (karlkfi) [reply e2e/testcases/stress_test.go:311]: fixed to 1000.

File: e2e/nomostest/artifactregistry/crane.go (hunk 5 above)
Comment (nan-yu) [reply e2e/nomostest/artifactregistry/crane.go:51]: ACK

File: e2e/nomostest/iam/account.go (hunk 3 above)
Comment (nan-yu) [reply e2e/nomostest/iam/account.go:26]: ACK

**Review by nan-yu — APPROVED** [review]
/lgtm
